sematic/types/types/union.py

```diff
@@ -50,8 +52,28 @@ def _union_can_cast(from_type: Any, to_type: Any) -> Tuple[bool, Optional[str]]:
 @register_to_json_encodable(Union)
 def _union_to_json_encodable(value: Any, type_: Any) -> Any:
     # We assume that casting has already vetted type_
-
-    return value_to_json_encodable(value, _get_value_type(value, type_))
+    value_type = _get_value_type(value, type_)
+    value_type_index = -1
+    for i, unioned_type in enumerate(get_args(type_)):
+        if unioned_type == value_type:
+            value_type_index = i
+            break
+    return {
+        "value": value_to_json_encodable(value, value_type),
+        "value_type_index": value_type_index,
```

Comment:
This is a breaking change that will make all already-persisted values unusable.

Also, this means ever reordering the Union types will make all already-persisted values unusable again.

Instead of specifying the index, and imposing this custom serialization schema, why not just serialize the value as before and add a deserialization method that tries all Union types in order to deserialize it? It is guaranteed that the same type that was used to cast and can-cast the value will be used, as cast and can-cast also iterate through the types in order.

Reply:
Makes sense, thank you!